src/transform/src/use_indexes.rs

```diff
@@ -63,283 +75,60 @@ impl FilterEqualLiteral {
+                        let best_index = key_set
                             .iter()
-                            .filter_map(|e| match e {
-                                e if *e == position => None,
-                                e if *e > position => Some(e - 1),
-                                _ => Some(*e),
+                            .filter(|ks| {
+                                ks.iter()
+                                    .all(|k| expr_to_equivalent_literal.contains_key(k))
                             })
-                            .collect();
-                    }
-                }
-                // if there is a cycle in the topological sort, don't bother sorting
-                //the rest of the variables
-                new_variables.append(variables);
-                *variables = new_variables;
-                // TODO(wangandi): considering turning the filters into wrappers instead of lifting them
-                // reorder the variables to match the key + convert input to arrangement if the variable
-                // rearrangement is a success
-                let input_types = inputs.iter().map(|i| i.typ()).collect::<Vec<_>>();
-                let input_arities = input_types
-                    .iter()
-                    .map(|i| i.column_types.len())
-                    .collect::<Vec<_>>();
-                let mut offset = 0;
-                let mut prior_arities = Vec::new();
-                for input in 0..inputs.len() {
-                    prior_arities.push(offset);
-                    offset += input_arities[input];
-                }
-                let mut lifted_predicates = Vec::new();
-                for (input_num, index) in matching_index_by_input {
-                    // if there was a cycle in the topological sort, test that the index can be used
-                    // in this variable ordering.
-                    if !incoming_edges.is_empty() {
-                        let mut index_iter = index.iter();
-                        let mut current_key = index_iter.next();
-                        for variable in variables.iter() {
-                            if let Some(key) = current_key {
-                                if let ScalarExpr::Column(c) = key {
-                                    if variable.contains(&(input_num, *c)) {
-                                        current_key = index_iter.next();
-                                    }
-                                }
-                            } else {
-                                break;
+                            .max_by_key(|ks| ks.len());
+                        if let Some(keys) = best_index {
+                            for key in keys {
+                                equivalences.push(vec![
+                                    key.clone(),
+                                    expr_to_equivalent_literal.remove(&key).unwrap(),
+                                ]);
```

Comment:
This may be minor, but does this handle cases where the same column reference occurs multiple times? Like, if I write (badly) `where col1 = 7 and col1 = 5`, I think this may just have one of the tests?

Reply:
I do have a test for the same column reference occurring multiple times in PR #2164, and it appears that no externally visible bad effects occur from this transform not handling the case, especially since `FilterEqualLiteral` does not remove the original filter that was converted to a join.
https://github.com/MaterializeInc/materialize/blob/9b2dfbfcecf1c7bfc6477d24bd7914332b6740b5/test/sqllogictest/transform/filter_index.slt#L359

I suspect that in the case of `col1=7` and `col1=5`, `ColumnKnowledge` or some other transform automatically converts the filters to `false` and then converts the entire input to a blank constant.

If you like, I can add code in this transform to handle the degenerate same column reference occurring multiple times case for future proofing sake.